Use Eager Code To Determine Conv Layout

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #87641
* #87650
* __->__ #87305

The logic for determine conv backend and therefore output striding is very complex. It depends on build settings, input striding/contiguity, sizes, etc. Eventually we should port that logic to the meta impl for dynamic shapes but that will require a lot more work and keeping the implementations in sync. See https://github.com/pytorch/torchdynamo/issues/1701

This is a prerequisite to removing the inductor conv stride propagation and more general fake tensor for inductor propagation. In that PR, the meta impls for cpu conv give incorrect striding which led to test failures (https://github.com/pytorch/pytorch/pull/87083).

eellison committed:
Fake Tensor Conv Fixes

[ghstack-poisoned]

eellison committed:
Update on "Use Eager Code To Determine Conv Layout"


The logic for determine conv backend and therefore output striding is very complex. It depends on build settings, input striding/contiguity, sizes, etc. Eventually we should port that logic to the meta impl for dynamic shapes but that will require a lot more work and keeping the implementations in sync. See https://github.com/pytorch/torchdynamo/issues/1701

This is a prerequisite to removing the inductor conv stride propagation and more general fake tensor for inductor propagation. In that PR, the meta impls for cpu conv give incorrect striding which led to test failures (https://github.com/pytorch/pytorch/pull/87083).  


[ghstack-poisoned]

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -422,11 +422,53 @@ def nyi(fake_mode, func, *args, **kwargs):
+            )
+        mem_fmt = torch._C._conv_determine_backend_memory_format(
+            kwargs["input"], kwargs["weight"], conv_backend
+        )
```
You don't actually care about the backend, right? Just the output memory format. Maybe only expose that?

eellison replied:
yea, `_select_conv_backend` was already exposed fwiw

bdhirsh commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -422,11 +422,53 @@ def nyi(fake_mode, func, *args, **kwargs):
-# FakeTensor extends MetaTensors to also carry an additional `fake_device`
-# which tracks devices that would have been used.
+@register_op_impl(
+    lambda func: func in (aten.convolution.default, aten.convolution_backward.default)
```
So the idea is that the conv meta impl in `_meta_registrations.py` is mostly correct, but gives the wrong memory format - so we manually call it and then patch up the memory format here in `FakeTensorMode`?

Is there any reason not to just patch the fancier memory format logic directly into the existing meta function? (If the answer is just that's more effort and work for later then totally fair - just want to confirm my understanding).

eellison replied:
Most of the complexity of the logic comes in cpu, so it's not that high pri at the moment, but it was a good way to unblock turning on fake tensors for inductor shape prop.

We will eventually need to patch the fancier memory format logic yea - but there might be a way to do something similar as we're doing here even for symbolic shapes

ezyang replied:
This logic should be moved into `_meta_registrations.py`

eellison replied:
Yea, I did a best effort fix [here](https://github.com/pytorch/torchdynamo/pull/1639). There is a [lot](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Convolution.cpp#L193) [lot](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Convolution.cpp#L239-L273) [lot](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Convolution.cpp#L313-L325) of stuff to copy over.

eellison committed:
Update on "Use Eager Code To Determine Conv Layout"


The logic for determine conv backend and therefore output striding is very complex. It depends on build settings, input striding/contiguity, sizes, etc. Eventually we should port that logic to the meta impl for dynamic shapes but that will require a lot more work and keeping the implementations in sync. See https://github.com/pytorch/torchdynamo/issues/1701

This is a prerequisite to removing the inductor conv stride propagation and more general fake tensor for inductor propagation. In that PR, the meta impls for cpu conv give incorrect striding which led to test failures (https://github.com/pytorch/pytorch/pull/87083).  


[ghstack-poisoned]

eellison committed:
Update on "Use Eager Code To Determine Conv Layout"


The logic for determine conv backend and therefore output striding is very complex. It depends on build settings, input striding/contiguity, sizes, etc. Eventually we should port that logic to the meta impl for dynamic shapes but that will require a lot more work and keeping the implementations in sync. See https://github.com/pytorch/torchdynamo/issues/1701

This is a prerequisite to removing the inductor conv stride propagation and more general fake tensor for inductor propagation. In that PR, the meta impls for cpu conv give incorrect striding which led to test failures (https://github.com/pytorch/pytorch/pull/87083).  


[ghstack-poisoned]

eellison committed:
Update on "Use Eager Code To Determine Conv Layout"


The logic for determine conv backend and therefore output striding is very complex. It depends on build settings, input striding/contiguity, sizes, etc. Eventually we should port that logic to the meta impl for dynamic shapes but that will require a lot more work and keeping the implementations in sync. See https://github.com/pytorch/torchdynamo/issues/1701

This is a prerequisite to removing the inductor conv stride propagation and more general fake tensor for inductor propagation. In that PR, the meta impls for cpu conv give incorrect striding which led to test failures (https://github.com/pytorch/pytorch/pull/87083).  


[ghstack-poisoned]

eellison committed:
Update on "Use Eager Code To Determine Conv Layout"


The logic for determine conv backend and therefore output striding is very complex. It depends on build settings, input striding/contiguity, sizes, etc. Eventually we should port that logic to the meta impl for dynamic shapes but that will require a lot more work and keeping the implementations in sync. See https://github.com/pytorch/torchdynamo/issues/1701

This is a prerequisite to removing the inductor conv stride propagation and more general fake tensor for inductor propagation. In that PR, the meta impls for cpu conv give incorrect striding which led to test failures (https://github.com/pytorch/pytorch/pull/87083).  


[ghstack-poisoned]

eellison committed:
Update on "Use Eager Code To Determine Conv Layout"


The logic for determine conv backend and therefore output striding is very complex. It depends on build settings, input striding/contiguity, sizes, etc. Eventually we should port that logic to the meta impl for dynamic shapes but that will require a lot more work and keeping the implementations in sync. See https://github.com/pytorch/torchdynamo/issues/1701

This is a prerequisite to removing the inductor conv stride propagation and more general fake tensor for inductor propagation. In that PR, the meta impls for cpu conv give incorrect striding which led to test failures (https://github.com/pytorch/pytorch/pull/87083).  


[ghstack-poisoned]

eellison committed:
Update on "Use Eager Code To Determine Conv Layout"


The logic for determine conv backend and therefore output striding is very complex. It depends on build settings, input striding/contiguity, sizes, etc. Eventually we should port that logic to the meta impl for dynamic shapes but that will require a lot more work and keeping the implementations in sync. See https://github.com/pytorch/torchdynamo/issues/1701

This is a prerequisite to removing the inductor conv stride propagation and more general fake tensor for inductor propagation. In that PR, the meta impls for cpu conv give incorrect striding which led to test failures (https://github.com/pytorch/pytorch/pull/87083).  


[ghstack-poisoned]

eellison committed:
Update on "Use Eager Code To Determine Conv Layout"


The logic for determine conv backend and therefore output striding is very complex. It depends on build settings, input striding/contiguity, sizes, etc. Eventually we should port that logic to the meta impl for dynamic shapes but that will require a lot more work and keeping the implementations in sync. See https://github.com/pytorch/torchdynamo/issues/1701

This is a prerequisite to removing the inductor conv stride propagation and more general fake tensor for inductor propagation. In that PR, the meta impls for cpu conv give incorrect striding which led to test failures (https://github.com/pytorch/pytorch/pull/87083).  


[ghstack-poisoned]

eellison committed:
Update on "Use Eager Code To Determine Conv Layout"


The logic for determine conv backend and therefore output striding is very complex. It depends on build settings, input striding/contiguity, sizes, etc. Eventually we should port that logic to the meta impl for dynamic shapes but that will require a lot more work and keeping the implementations in sync. See https://github.com/pytorch/torchdynamo/issues/1701

This is a prerequisite to removing the inductor conv stride propagation and more general fake tensor for inductor propagation. In that PR, the meta impls for cpu conv give incorrect striding which led to test failures (https://github.com/pytorch/pytorch/pull/87083).  


[ghstack-poisoned]

eellison committed:
Update on "Use Eager Code To Determine Conv Layout"


The logic for determine conv backend and therefore output striding is very complex. It depends on build settings, input striding/contiguity, sizes, etc. Eventually we should port that logic to the meta impl for dynamic shapes but that will require a lot more work and keeping the implementations in sync. See https://github.com/pytorch/torchdynamo/issues/1701

This is a prerequisite to removing the inductor conv stride propagation and more general fake tensor for inductor propagation. In that PR, the meta impls for cpu conv give incorrect striding which led to test failures (https://github.com/pytorch/pytorch/pull/87083).  


[ghstack-poisoned]

eellison commented:
@pytorchbot merge
